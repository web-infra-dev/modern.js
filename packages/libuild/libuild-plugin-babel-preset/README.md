# @modern-js/libuild-plugin-babel-preset
[![npm version](https://badge.fury.io/js/@modern-js%2Flibuild-plugin-babel-preset.svg)](https://www.npmjs.com/package/@modern-js/libuild-plugin-babel-preset)

A plugin for libuild to transform your code with @babel/preset-react, @babel/preset-typescript and babel-plugin-import

## Tips

For compatibility with older applications, Libuild also supports user-defined Babel configurations, such as specifying additional babel plugins and babel presets, but please note: **Using Babel will result in slower builds, so do not use Babel unless necessary**. 

Neither `.babelrc` nor `babel.config.js` in the project root will take effect. If you need to use Babel, you need to configure it in the `libuild.config.ts` file.

## Usage

```ts
// libuild.config.ts
import { defineConfig } from '@modern-js/libuild';
import { babelPresetPlugin } from '@modern-js/libuild-plugin-babel-preset';

export = defineConfig({
  plugins: [
    babelPresetPlugin({
      import: [
        {
          libraryName: 'antd',
          style: false,
        }
        // babel-plugin-import options
      ], 
      react: {} // @babel/preset-react options
    }, {
      // babel-core transform options
    })
  ]
})
```

## Options

### preset options
`react`: [babel-preset-react](https://babeljs.io/docs/en/babel-preset-react#options)

`import`: [babel-plugin-import](https://github.com/umijs/babel-plugin-import#options)

### transform options
See here for [babel transform options](https://babeljs.io/docs/en/options)