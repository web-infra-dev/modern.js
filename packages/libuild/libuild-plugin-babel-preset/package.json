{
  "name": "@modern-js/libuild-plugin-babel-preset",
  "version": "0.14.3",
  "description": "A plugin for libuild to transform your code with @babel/preset-react, @babel/preset-typescript and babel-plugin-import",
  "keywords": [
    "libuild",
    "babel",
    "preset",
    "react",
    "typescript",
    "import"
  ],
  "repository": "modern-js-dev/libuild",
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "watch": "libuild -w",
    "build": "tsc && libuild"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "eslint": "7.32.0",
    "@modern-js/eslint-config": "workspace:*",
    "@modern-js/libuild": "workspace:*",
    "@modern-js/libuild-utils": "workspace:*",
    "typescript": "4.4.4",
    "@types/babel__core": "7.1.16"
  },
  "dependencies": {
    "babel-plugin-import": "1.13.3",
    "@babel/preset-react": "7.14.5",
    "@babel/preset-typescript": "7.16.7",
    "@babel/core": "7.15.8"
  },
  "files": [
    "dist",
    "!dist/*.map"
  ]
}
