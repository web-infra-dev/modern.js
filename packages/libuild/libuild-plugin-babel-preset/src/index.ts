import type { LibuildPlugin } from '@modern-js/libuild';
import type { TransformOptions as BabelTransformOptions } from '@babel/core';
import { isJsExt, isJsLoader, resolvePathAndQuery } from '@modern-js/libuild-utils';

export interface BabelPluginImportOptions {
  libraryName: string;
  /**
   * @default "lib"
   */
  libraryDirectory?: string;

  /**
   * @default true
   */
  camel2DashComponentName?: boolean;

  style: boolean | ((name: string) => string);

  styleLibraryDirectory?: string;

  customName?: string;

  customStyleName?: string;

  transformToDefaultImport?: boolean;
}
export interface BabelReactOptions {
  /**
   * @default "React.createElement" (only in classic runtime)
   */
  pragma?: string;
  /**
   * @default "React.Fragment" (only in classic runtime)
   */
  pragmaFrag?: string;
  /**
   * @default true
   */
  throwIfNamespace?: boolean;
  /**
   * @default "classic"
   */
  runtime?: 'classic' | 'automatic';

  /**
   * @default "react"  (only in automatic runtime)
   */
  importSource?: string;
}

export interface BabelPresetOption {
  /**
   * @description babel-plugin-import options
   */
  import?: BabelPluginImportOptions | BabelPluginImportOptions[];
  /**
   * @description @babel/preset-react options.
   */
  react?: BabelReactOptions;
}

export const babelPresetPlugin = (
  preset: BabelPresetOption = {},
  options: BabelTransformOptions = {}
): LibuildPlugin => {
  return {
    name: 'libuild:babel-preset',
    apply(compiler) {
      compiler.hooks.transform.tapPromise('babel', async (args) => {
        const { originalFilePath } = resolvePathAndQuery(args.path);
        if (isJsExt(originalFilePath) || isJsLoader(args.loader)) {
          const isTsx = args.loader === 'tsx' || /\.tsx$/i.test(originalFilePath);
          const { plugins, presets } = normalizeBabel(preset, isTsx);
          const result = await require('@babel/core').transformAsync(args.code, {
            ...options,
            exclude: [/\bcore-js\b/],
            inputSourceMap: false,
            filename: originalFilePath,
            sourceFileName: originalFilePath,
            sourceMaps: Boolean(compiler.config.sourceMap),
            babelrc: false,
            configFile: false,
            compact: false,
            presets: [...presets, ...(options.presets || [])],
            plugins: [...plugins, ...(options.plugins || [])],
          });

          return {
            ...args,
            code: result?.code,
            map: result?.map,
          };
        }
        /* c8 ignore start */
        return args;
        /* c8 ignore stop */
      });
    },
  };
};

function normalizeBabel(preset: BabelPresetOption, isTsx: boolean) {
  const presets = [[require('@babel/preset-typescript'), isTsx ? { isTSX: true, allExtensions: true } : {}]];

  /* c8 ignore start */
  if (preset.react || preset.import) {
    presets.push([require('@babel/preset-react'), preset.react]);
  }

  const plugins = [];

  if (preset.import) {
    if (Array.isArray(preset.import)) {
      plugins.push(
        ...preset.import.map((o, i) => {
          return [require('babel-plugin-import'), o, String(i)];
        })
      );
    } else if (typeof preset.import === 'object') {
      plugins.push([require('babel-plugin-import'), preset.import]);
    }
  }
  /* c8 ignore stop */
  return { presets, plugins };
}
