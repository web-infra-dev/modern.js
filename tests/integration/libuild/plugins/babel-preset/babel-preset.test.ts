import { getLibuilderTest } from '@/toolkit';
import { babelPresetPlugin } from '@modern-js/libuild-plugin-babel-preset';

describe('fixture:plugin:babel-preset', function () {
  it('plugin:babel-preset:options:react', async () => {
    const bundler = await getLibuilderTest({
      root: __dirname,
      input: {
        main: './index.tsx',
        ts: '/index.ts?virtual',
      },
      plugins: [
        babelPresetPlugin({
          react: {},
        }),
      ],
    });
    await bundler.build();

    bundler.expectJSOutputMatchSnapshot();
  });

  it('plugin:babel-preset:options:import', async () => {
    const bundler = await getLibuilderTest({
      root: __dirname,
      input: {
        main: './index.tsx',
        ts: '/index.ts?virtual',
      },
      plugins: [
        babelPresetPlugin({
          import: {
            libraryName: 'antd',
            style: false,
          },
        }),
      ],
    });
    await bundler.build();

    bundler.expectJSOutputMatchSnapshot();
  });

  it('plugin:babel-preset:options:import-array', async () => {
    const bundler = await getLibuilderTest({
      root: __dirname,
      input: {
        main: './index.tsx',
        ts: '/index.ts?virtual',
      },
      plugins: [
        babelPresetPlugin({
          import: [
            {
              libraryName: 'antd',
              style: false,
            },
            {
              libraryName: 'antd',
              style: false,
            },
          ],
        }),
      ],
    });
    await bundler.build();

    bundler.expectJSOutputMatchSnapshot();
  });
});
