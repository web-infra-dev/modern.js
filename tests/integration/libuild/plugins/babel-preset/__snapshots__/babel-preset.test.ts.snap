// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixture:plugin:babel-preset plugin:babel-preset:options:import 1`] = `
"// index.tsx
import _Button from \\"antd/lib/button\\";
var value = /* @__PURE__ */ React.createElement(_Button, null);
console.log(value);
"
`;

exports[`fixture:plugin:babel-preset plugin:babel-preset:options:import 2`] = `
"// libuild:virtual:/index.ts?virtual
import _Button from \\"antd/lib/button\\";
var value = /* @__PURE__ */ React.createElement(_Button, null);
console.log(value);
"
`;

exports[`fixture:plugin:babel-preset plugin:babel-preset:options:import-array 1`] = `
"// index.tsx
import _Button from \\"antd/lib/button\\";
var value = /* @__PURE__ */ React.createElement(_Button, null);
console.log(value);
"
`;

exports[`fixture:plugin:babel-preset plugin:babel-preset:options:import-array 2`] = `
"// libuild:virtual:/index.ts?virtual
import _Button from \\"antd/lib/button\\";
var value = /* @__PURE__ */ React.createElement(_Button, null);
console.log(value);
"
`;

exports[`fixture:plugin:babel-preset plugin:babel-preset:options:react 1`] = `
"// index.tsx
import { Button } from \\"antd\\";
var value = /* @__PURE__ */ React.createElement(Button, null);
console.log(value);
"
`;

exports[`fixture:plugin:babel-preset plugin:babel-preset:options:react 2`] = `
"// libuild:virtual:/index.ts?virtual
import { Button } from \\"antd\\";
var value = /* @__PURE__ */ React.createElement(Button, null);
console.log(value);
"
`;
